src/AutoRest.CSharp/Common/Generation/Writers/CodeWriterExtensions.cs

```diff
@@ -92,74 +93,45 @@ public static CodeWriter WriteFieldDeclaration(this CodeWriter writer, FieldDecl
 
         public static CodeWriter.CodeWriterScope WriteMethodDeclaration(this CodeWriter writer, MethodSignatureBase methodBase, params string[] disabledWarnings)
         {
-            WriteDisableWarnings(writer, disabledWarnings);
-
-            writer.Append($"{methodBase.Modifiers} ");
-            if (methodBase is MethodSignature method)
+            foreach (var disabledWarning in disabledWarnings)
             {
-                if (method.ReturnType != null)
-                {
-                    writer.Append($"{method.ReturnType} ");
-                }
-                else
-                {
-                    writer.AppendRaw("void ");
-                }
+                writer.Line($"#pragma warning disable {disabledWarning}");
             }
 
-            return WriteMethodDeclarationParameters(writer, methodBase, disabledWarnings, methodBase.Name);
-        }
+            writer
+                .AppendRawIf("public ", methodBase.Modifiers.HasFlag(Public))
+                .AppendRawIf("internal ", methodBase.Modifiers.HasFlag(Internal))
+                .AppendRawIf("protected ", methodBase.Modifiers.HasFlag(Protected))
+                .AppendRawIf("private ", methodBase.Modifiers.HasFlag(Private));
 
-        public static CodeWriter.CodeWriterScope WriteMethodDeclaration(this CodeWriter writer, MethodSignatureBase methodBase, bool isAsync, params string[] disabledWarnings)
-        {
-            WriteDisableWarnings(writer, disabledWarnings);
 
-            writer.Append($"{methodBase.Modifiers} ");
             if (methodBase is MethodSignature method)
             {
-                if (isAsync && !method.IsPageable)
-                    writer.Append($"async ");
-
-                var firstParam = method.Parameters.FirstOrDefault();
-                bool isExtensionMethod = firstParam is not null && firstParam.IsExtensionParameter;
-
-                if (method.Modifiers.Contains("public") && !isExtensionMethod)
-                    writer.Append($"virtual ");
-
-                if (isExtensionMethod)
-                    writer.Append($"static ");
+                writer
+                    .AppendRawIf("async ", methodBase.Modifiers.HasFlag(Async) && Configuration.AzureArm)
```

Comment:
ARM has been writing `async` before `virtual`. Canonical order is when `async` goes after `virtual`, but it can be fixed in a separate PR.